Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerColorAnnotator.java

```diff
@@ -0,0 +1,147 @@
+    }
+    if (type.equals(HighlightRegionType.GETTER_DECLARATION)) {
+      return DartSyntaxHighlighterColors.DART_GETTER_DECLARATION;
+    }
+    if (type.equals(HighlightRegionType.LOCAL_VARIABLE)) {
+      return DartSyntaxHighlighterColors.DART_LOCAL_VARIABLE_ACCESS;
+    }
+    if (type.equals(HighlightRegionType.IMPORT_PREFIX)) {
+      return DartSyntaxHighlighterColors.DART_IMPORT_PREFIX;
+    }
+    if (type.equals(HighlightRegionType.KEYWORD)) {
+      return DartSyntaxHighlighterColors.DART_KEYWORD;
+    }
+    if (type.equals(HighlightRegionType.LABEL)) {
+      return DartSyntaxHighlighterColors.DART_LABEL;
+    }
+    if (type.equals(HighlightRegionType.LOCAL_VARIABLE_DECLARATION)) {
+      return DartSyntaxHighlighterColors.DART_LOCAL_VARIABLE;
+    }
+    if (type.equals(HighlightRegionType.METHOD)) {
+      return DartSyntaxHighlighterColors.DART_INSTANCE_MEMBER_FUNCTION_CALL;
+    }
+    if (type.equals(HighlightRegionType.METHOD_STATIC)) {
+      return DartSyntaxHighlighterColors.DART_STATIC_MEMBER_FUNCTION_CALL;
+    }
+    if (type.equals(HighlightRegionType.METHOD_DECLARATION)) {
+      return DartSyntaxHighlighterColors.DART_INSTANCE_MEMBER_FUNCTION;
+    }
+    if (type.equals(HighlightRegionType.METHOD_DECLARATION_STATIC)) {
+      return DartSyntaxHighlighterColors.DART_STATIC_MEMBER_FUNCTION;
+    }
+    if (type.equals(HighlightRegionType.PARAMETER)) {
+      return DartSyntaxHighlighterColors.DART_PARAMETER;
+    }
+    if (type.equals(HighlightRegionType.SETTER_DECLARATION)) {
+      return DartSyntaxHighlighterColors.DART_SETTER_DECLARATION;
+    }
+    if (type.equals(HighlightRegionType.TOP_LEVEL_VARIABLE)) {
+      return DartSyntaxHighlighterColors.DART_TOP_LEVEL_VARIABLE_DECLARATION;
+    }
+    if (type.equals(HighlightRegionType.TYPE_NAME_DYNAMIC)) {
+      return DartSyntaxHighlighterColors.DART_KEYWORD;
+    }
+    if (type.equals(HighlightRegionType.TYPE_PARAMETER)) {
+      return DartSyntaxHighlighterColors.DART_TYPE_PARAMETER;
+    }
+    return null;
+  }
+
+  @Override
+  public void apply(@NotNull PsiFile file, Object annotationResult, @NotNull AnnotationHolder holder) {
+    final VirtualFile virtualFile = file.getVirtualFile();
+    final List<PluginHighlightRegion> regions = DartAnalysisServerService.getInstance().getHighlight(virtualFile);
+    for (PluginHighlightRegion region : regions) {
+      final String type = region.getType();
+      final String key = getHighlightType(type);
+      if (key != null) {
+        final TextAttributesKey textAttributes = TextAttributesKey.createTextAttributesKey(key);
+        final TextRange textRange = new TextRange(region.getOffset(), region.getOffset() + region.getLength());
+        holder.createInfoAnnotation(textRange, null).setTextAttributes(textAttributes);
+      }
+    }
+  }
+
+  @Nullable
+  @Override
+  public Object collectInformation(@NotNull PsiFile file, @NotNull Editor editor, boolean hasErrors) {
```

Comment:
Anyway it is out of scope of this pull request. Just to think about.
I like how DartAnalysisServerAnnotator works.
It sends updated content in collectInformation().
It asks and waits for errors in doAnnotate(). When errors are received we are sure that they are actual at the moment.
Then it applies errors in apply().

Semantic highlights are different as they are not returned based on client request, but notify listener instead. So I'm thinking how to mimic behavior of server.analysis_getErrors() that has a callback.

Not for this pull request anyway. I'm pushing it.